File: Workshops/WS4.py
# Qustion 1
class Student:
    # Class attribute
    school_name = "WE-School"
    # Instance attributes
    def __init__(self, name, grade):
        self.name = name
        self.grade = grade
    # Method to display student details
    def display_info(self):
        print(f"Student Name: {self.name}")
        print(f"Grade: {self.grade}")
        print(f"School: {Student.school_name}")
        
# Creating objects of the Student class
student1 = Student("Ahmed", "2nd")
student2 = Student("Ali", "3rd")

# Calling methods on those objects
student1.display_info()
student2.display_info()

# Question 2
class Book:
    """Constructor with instance attributes"""
    def __init__(self, title, author):
        self.title = title
        self.author = author
        self.is_borrowed = False # Initially, the book is not borrowed
    """Method to borrow the book"""
    def borrow_book(self):
        if not self.is_borrowed:
            self.is_borrowed = True
            print(f"{self.title} by {self.author} has been borrowed.")
        else:
            print(f"{self.title} is currently not available.")
    """Method to return the book"""
    def return_book(self):
        if self.is_borrowed:
            self.is_borrowed = False
            print(f"{self.title} has been returned.")
        else:
            print(f"{self.title} was not borrowed.")

# Creating a Book object
book1 = Book("Python Programming", "Ahmed Ali")
# Borrowing and returning the book
book1.borrow_book()
book1.return_book()

File: Workshops/WS4.js
// Qustion 1
class Student {
    // Class attribute
    static get school_name() {
        return "WE-School";
    }
    // Constructor for instance attributes
    constructor(name, grade) {
        this.name = name;
        this.grade = grade;
    }
    // Method to display student details
    displayInfo() {
        console.log(`Student Name: ${this.name}`);
        console.log(`Grade: ${this.grade}`);
        console.log(`School: ${Student.school_name}`);
    }
}
// Creating objects of the Student class
let student1 = new Student("Ahmed", "2nd");
let student2 = new Student("Ali", "3rd");
// Calling methods on those objects
student1.displayInfo();
student2.displayInfo();

// Qustion 2
class Book {
    constructor(title, author) {
        this.title = title;
        this.author = author;
        this.isBorrowed = false; // Initially, the book is not borrowed
    }
    /* Method to borrow the book */
    borrowBook() {
        if (!this.isBorrowed) {
            this.isBorrowed = true;
            console.log(`${this.title} by ${this.author} has been borrowed.`);
        } else {
            console.log(`${this.title} is currently not available.`);
        }
    }
    /** Method to return the book */
    returnBook() {
        if (this.isBorrowed) {
            this.isBorrowed = false;
            console.log(`${this.title} has been returned.`);
        } else {
            console.log(`${this.title} was not borrowed.`);
        }
    }
}
// Creating a Book object
const book1 = new Book("Python Programming", "Ahmed Ali");
// Borrowing and returning the book
book1.borrowBook();
book1.returnBook();